### PR TITLE
Reolink battery params supported

### PR DESCRIPTION
**Pull request by apocaliss92**

### CODE REVIEW

File: plugins/reolink/src/main.ts [hunk 1]
```diff
@@ -318,10 +324,28 @@ class ReolinkCamera extends RtspSmartCamera implements Camera, DeviceProvider, R
...
+        if (this.batteryTimeout) {
+            clearInterval(this.batteryTimeout);
+        }
+
+        this.batteryTimeout = setInterval(async () => {
+            const api = this.getClientWithToken();
+
+            const { batteryPercent, sleep } = await api.getBatteryInfo();
+            this.batteryLevel = batteryPercent;
+            this.online = !sleep;
```
Comment (apocaliss92) [inline plugins/reolink/src/main.ts:345]: This sleep parameter is a very useful thing to keep to build more stuff later on, not sure if it's the right place where to store it? It tells if the camera is in standby. An use case I use it for is to generate a snapshot as soon as it comes online to save battery even more

File: plugins/reolink/src/main.ts [hunk 2]
```diff
@@ -222,7 +223,7 @@ class ReolinkCamera extends RtspSmartCamera implements Camera, DeviceProvider, R
         let abilities;
         try {
             abilities = await api.getAbility();
-        } catch(e) {
```
Comment (koush) [inline plugins/reolink/src/main.ts:225]: undo

Comment (apocaliss92) [reply plugins/reolink/src/main.ts]: done in both lines

File: plugins/reolink/src/main.ts [hunk 3]
```diff
@@ -293,6 +294,11 @@ class ReolinkCamera extends RtspSmartCamera implements Camera, DeviceProvider, R
             && this.storageSettings.values.abilities?.value?.Ability?.supportAudioAlarm?.ver !== 0;
     }
 
+    hasBattery() {
+        const batteryConfigVer = this.storageSettings.values.abilities?.value?.Ability?.abilityChn?.[this.getRtspChannel()].battery?.ver ?? 0;
```
Comment (koush) [inline plugins/reolink/src/main.ts:298]: need optional index.

        const batteryConfigVer = this.storageSettings.values.abilities?.value?.Ability?.abilityChn?.[this.getRtspChannel()]?.battery?.ver ?? 0;

File: plugins/reolink/src/main.ts [hunk 4]
```diff
@@ -318,10 +324,28 @@ class ReolinkCamera extends RtspSmartCamera implements Camera, DeviceProvider, R
...
             interfaces.push(ScryptedInterface.DeviceProvider);
+        if (this.hasBattery()) {
+            interfaces.push(ScryptedInterface.Battery, ScryptedInterface.Online);
+            await this.startBatteryCheckInterval();
+        }
 
         await this.provider.updateDevice(this.nativeId, name, interfaces, type);
     }
 
+    async startBatteryCheckInterval() {
```
Comment (koush) [inline plugins/reolink/src/main.ts:335]: I dont think this function needs to be async?

File: plugins/reolink/src/main.ts [hunk 5]
```diff
@@ -318,10 +324,28 @@ class ReolinkCamera extends RtspSmartCamera implements Camera, DeviceProvider, R
         }
         if (this.hasSiren())
             interfaces.push(ScryptedInterface.DeviceProvider);
+        if (this.hasBattery()) {
+            interfaces.push(ScryptedInterface.Battery, ScryptedInterface.Online);
+            await this.startBatteryCheckInterval();
```
Comment (koush) [inline plugins/reolink/src/main.ts:329]: await not necessary, see comment below.

File: plugins/reolink/src/main.ts [hunk 6]
```diff
@@ -318,10 +324,28 @@ class ReolinkCamera extends RtspSmartCamera implements Camera, DeviceProvider, R
...
 
+    async startBatteryCheckInterval() {
+        if (this.batteryTimeout) {
+            clearInterval(this.batteryTimeout);
+        }
+
+        this.batteryTimeout = setInterval(async () => {
+            const api = this.getClientWithToken();
+
+            const { batteryPercent, sleep } = await api.getBatteryInfo();
```
Comment (koush) [inline plugins/reolink/src/main.ts:343]: need failure handling.

File: plugins/reolink/src/main.ts (hunk 3 above)
Comment (apocaliss92) [reply plugins/reolink/src/main.ts]: wops, done

File: plugins/reolink/src/main.ts (hunk 4 above)
Comment (apocaliss92) [reply plugins/reolink/src/main.ts]: removed

File: plugins/reolink/src/main.ts (hunk 5 above)
Comment (apocaliss92) [reply plugins/reolink/src/main.ts]: removed

File: plugins/reolink/src/main.ts (hunk 6 above)
Comment (apocaliss92) [reply plugins/reolink/src/main.ts]: added

**Comment by koush** [comment]
why close? haven't had time to review yet

**Pull request merged**